internal/provider/resource_workspace_current_state.go

```diff
@@ -0,0 +1,313 @@
+		Description:         description,
+		Attributes: map[string]schema.Attribute{
+			"workspace_path": schema.StringAttribute{
+				MarkdownDescription: "The full path of the workspace.",
+				Description:         "The full path of the workspace.",
+				Required:            true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+			"module_source": schema.StringAttribute{
+				MarkdownDescription: "The source of the module, including the API hostname.",
+				Description:         "The source of the module, including the API hostname.",
+				Required:            true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+			"module_version": schema.StringAttribute{
+				MarkdownDescription: "The version identifier of the module.",
+				Description:         "The version identifier of the module.",
+				Required:            true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+			"teardown": schema.BoolAttribute{
+				MarkdownDescription: "Whether to teardown the deployment of the module to the workspace.",
+				Description:         "Whether to teardown the deployment of the module to the workspace.",
+				Optional:            true,
+				Computed:            true, // API sets a default value of false if not specified.
+				// Can be updated in place, so no RequiresReplace plan modifier.
+			},
+		},
+	}
+}
+
+// Configure lets the provider implement the ResourceWithConfigure interface.
+func (t *workspaceCurrentStateResource) Configure(_ context.Context,
+	req resource.ConfigureRequest, _ *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+	t.client = req.ProviderData.(*tharsis.Client)
+}
+
+func (t *workspaceCurrentStateResource) Create(ctx context.Context,
+	req resource.CreateRequest, resp *resource.CreateResponse) {
+
+	// Retrieve values from workspace current state.
+	var workspaceCurrentState WorkspaceCurrentStateModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &workspaceCurrentState)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// If teardown is not specified, make it false.
+	workspaceCurrentState.setDefaultTeardown()
+
+	if !workspaceCurrentState.Teardown.ValueBool() {
+		t.doApplyOrDestroyRun(ctx, workspaceCurrentState, resp.Diagnostics)
+	}
+
+	// Set the response state to the fully-populated plan, whether or not there is an error.
+	resp.Diagnostics.Append(resp.State.Set(ctx, workspaceCurrentState)...)
+}
+
+func (t *workspaceCurrentStateResource) Read(ctx context.Context,
+	req resource.ReadRequest, resp *resource.ReadResponse) {
+
+	// Get the current state.
+	var state WorkspaceCurrentStateModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
```

Comment:
Read will need to get the current workspace state to determine if the specified module and version are already deployed.

Reply:
If by "get the current workspace state", you mean to get the workspace object from Tharsis, there are no fields that tell us what was deployed.  Instead, I think I can do a GetRun based on the workspace path and check the module source and module version from there.

Reply:
Correction: I think I can do a GetRuns based on the workspace path, sorted by descending time of last update, and take the first one returned.